File: Observer/Go/observer/concreteSubject.go
package observer

type ConcreteSubject struct {
	value     string
	observers []Observer
}

func (cs *ConcreteSubject) RegisterObserver(ob Observer) {
	cs.observers = append(cs.observers, ob)
}

func (cs *ConcreteSubject) Notify() {
	for _, observer := range cs.observers {
		observer.Update(cs.value)
	}
}

func (cs *ConcreteSubject) SetValue(val string) {
	cs.value = val
	cs.Notify()
}

func (cs *ConcreteSubject) GetValue() string {
	return cs.value
}

File: Observer/PHP/concreteSubject.php
<?php
  class ConcreteSubject implements Subject
  {
    private $_observers, $_value;

    public function __construct()
    {
      $this->_observers = array();
    }

    public function registerObserver(Observer $observer)
    {
      array_push($this->_observers, $observer);
    }

    public function removeObserver(Observer $observer)
    {
      if ($index = array_search($observer, $this->_observers, true))
      {
        unset($this->_observers[$index]);
      }
    }

    public function notify()
    {
      foreach ($this->_observers as $observer)
      {
        $observer->update($this->_value);
      }
    }

    public function setValue($value)
    {
      $this->_value = $value;
      $this->notify();
    }

    public function getValue()
    {
      return $this->_value;
    }
  }
?>
